Feat/load average implementation

ariasmn opened the pull request:
This should close #120 

As I said on the issue, I had some problems for the `darwin` implementation, since I had to do it from a VM. I tested it without CGO (`CGO_ENABLED=0`) and still worked.

The way the load is parsed seems a bit tricky, although it's the standard way of doing so for BSD systems using Golang. I tried parsing the values the same way you did on others `SysctlRaw` without any luck (I'd appreciate an explanation if you know the reason, I'm still pretty new to Golang)

If anything needs some changes, please let me know!

Linux host: Fedora 36 Workstation, Kernel 5.19.15-201
Darwin host: macOS Catalina, Version 10.15.7

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-sysinfo-mbp%2FPR-140/detail/PR-140/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-sysinfo-mbp%2FPR-140/detail/PR-140/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-sysinfo-mbp%2FPR-140/detail/PR-140/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-sysinfo-mbp%2FPR-140/detail/PR-140/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-10-19T17:07:06.245Z&rangeTo=2022-10-19T17:27:06.245Z&transactionName=BUILD+Library%2Fgo-sysinfo-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=c264197c6dfaf940861a494d2c1c12b7&transactionId=f39781a5bc6a122e)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-10-19T17:17:06.245+0000


* Duration: 2 min 59 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-sysinfo-mbp%2FPR-140/detail/PR-140/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Load a resource file from a library`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Library/pipelines/go-sysinfo-mbp/pipelines/PR-140/runs/1/steps/81/log/?start=0">here</a></li>
<li>Description: <code>approval-list/elastic/go-sysinfo.yml</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Library/pipelines/go-sysinfo-mbp/pipelines/PR-140/runs/1/steps/89/log/?start=0">here</a></li>
<li>Description: <code>githubApiCall: The REST API call https://api.github.com/orgs/elastic/members/ariasmn return the message : java.lang.Exception: httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/ariasmn : httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/ariasmn : Code: 404Error: {"message":"User does not exist or is not a member of the organization","documentation_url":"https://docs.github.com/rest/reference/orgs#check-organization-membership-for-a-user"}</code></l1>
</ul>
  
  </p>
  </details>




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

prologic reviewed (changes requested):
Hey @ariasmn I tested this on my Mac and everything you've done here is working great 👌 -- Except for one minor patch/change you forgot to add 😅

```diff
commit 93d5e2259916ce787119d86ca91669125fbf6490
Author: James Mills <prologic@shortcircuit.net.au>
Date:   Thu Oct 20 08:43:26 2022 +1000

    Fix missing embedding of LoadAverage interface in Host interface

diff --git a/types/host.go b/types/host.go
index d2911ee..2ffee7b 100644
--- a/types/host.go
+++ b/types/host.go
@@ -24,6 +24,7 @@ import "time"
 // implementation is unable to collect all of the necessary data.
 type Host interface {
        CPUTimer
+       LoadAverage
        Info() HostInfo
        Memory() (*HostMemoryInfo, error)
 }
```

prologic commented:
I also put up a PR #41 that adds a little CLI that makes it a bit easier to test and demonstrate the package.

prologic commented:
No worries 🤗

andrewkroh commented on types/host.go:
```diff
@@ -98,7 +98,7 @@ type OSInfo struct {
 // LoadAverage returns load info on the host
 type LoadAverage interface {
-	LoadAverage() LoadAverageInfo
+	LoadAverage() (*LoadAverageInfo, error)
```
Please add a changelog entry in CHANGELOG.md and note that is a breaking API change.

prologic replied:
Technically this is new APi that didn't exist before that _should_ have. So this is not a breaking APi change IMO, it _should_ be backwards compatible. Yes the "interface" has changed, but that will only affect new projects? Even old projects would still compile AFAIK.

andrewkroh replied:
It is a breaking change. The interface already existed. Any users that had written code to use it will be broken by this change. It does not matter that there was no implementation, code that was written to expect `LoadAverage() LoadAverageInfo` won't compile now.

But given that there was no actual implementation I doubt that any code exists that actually tried to use it. And because of this, I'm OK with make this change to the interface. It would be a good idea to do a search on cs.github.com to see if we can find any usages.

prologic replied:
Yeah that was kind of my point, but you put it better 👌

ariasmn replied:
Added to the `CHANGELOG.md`.

Please, have a look, maybe the explanation is a bit odd since I'm not a native speaker.

andrewkroh reviewed (changes requested):
Can you please add invocation of `LoadAverage` to the system_test.go file.

prologic commented:
@ariasmn Can you see to the above? 🙏 I'd really like to see this PR merged 👌

ariasmn commented:
@prologic @andrewkroh Done!

Let me know if anything still needs some work.

andrewkroh commented:
/test

ariasmn commented:
Didn't even think about Windows now that I see the tests failing to be honest.

If I'm not mistaken, the only thing that I should do is just implement the `LoadAverage` function on the `host_windows.go`? Something like:

```
func (h *host) LoadAverage() (*types.LoadAverageInfo, error) {
	return nil, fmt.Errorf("LoadAverage is not available for this platform %w", types.ErrNotImplemented)
}
```

ariasmn commented:
I just added a commit to the PR, adding the missing functions on Windows and AIX hosts.
Also, I updated the README as well to note the new `LoadAverage` interface support.

Let me know if anything is still left (sorry that the PR is bouncing back and forth this much, completely my fault!)

andrewkroh commented on types/host.go:
```diff
@@ -25,6 +25,7 @@ import "time"
 type Host interface {
 	CPUTimer
 	Info() HostInfo
+	LoadAverage() (*LoadAverageInfo, error)
```
I think this should not be added to the `Host` interface. It's an optional thing given that platforms like Windows do not provide it. Callers that want to use it would check if it's available for their platform like

```
	if v, ok := host.(types.LoadAverage); ok {
		load, err := v.LoadAverage()
		if err != nil {
			return err
		}
		fmt.Printf("Load: %#v\n", load)
	}
```

Once you do this then the aix/windows implementations that return errors should go away.

ariasmn replied:
Done! To be honest it makes more sense that adding empty implementations on each host, thanks for the heads up.

Please, have a deep look at the `system_test.go` file. I don't really know if it's correct to add the invocation of `LoadAverage` to the `TestHost` function. If not, let me know the best place to add the invocation.

andrewkroh commented on system_test.go:
```diff
@@ -241,20 +241,27 @@ func TestHost(t *testing.T) {
-		t.Fatal(err)
+	if v, ok := host.(types.LoadAverage); ok {
+		loadAvg, err := v.LoadAverage()
+		if errors.Is(err, types.ErrNotImplemented) {
```
This is not needed given that not of the implementations return this error.

ariasmn replied:
You're right, didn't think about it, my bad!

Everything should be ok now.

andrewkroh commented:
/test

andrewkroh commented:
/test

andrewkroh commented:
/test

ariasmn commented:
@andrewkroh What's the status of this?
Just asking to know if there's anything left I should do on my side.
